client_side/.infrastructure/infrastructure.sh

```diff
@@ -0,0 +1,201 @@
+    echo "and Tellina <URL> to help you solve the tasks."
+  else
+    echo "but you can't use Tellina to help you solve the tasks."
+  fi
+}
+
+# Prints the current task number and its description.
+print_task() {
+  local task_code=$(cat "${INFRA_DIR}/.task_code")
+
+  echo "-----------------------------------------------------------------------"
+  echo "Task: ${curr_task}/${TASKS_SIZE}"
+
+  jq -r '.description' "${TASKS_DIR}/task_${task_code}/task_${task_code}.json"
+}
+
+# See documentation for ./verify_task.py for more details on what it does
+# Captures the stdout of verify_task.py into $status
+# Captures the exit code of verify_task.py into $EXIT
+verify_task() {
+  # Verify the output of the previous command.
+  local task_code="$(cat "${INFRA_DIR}/.task_code")"
+  local user_command="$(cat "${INFRA_DIR}/.command")"
+
+  status=$("${INFRA_DIR}"/verify_task.py ${task_code} ${user_command})
+  EXIT=$?
+}
+
+# Increment the current task number and update the true task code in .task_code
+# accordingly
+#   - Check if all the tasks are complete, if it is then skip the following
+#     steps and clean up instead.
+#   - If this current_task is equal to `TASKS_SIZE / 2`, switch treatments and
+#     task sets and notify the user.
+# Write "start task" to `.command`.
+# Writes to the log
+# Set time_elapsed=0, status="incomplete", SECONDS=0
+# Prints out the task's information
+next_task() {
+  make_fs
+
+  # Increment the number of tasks finished by the user
+  curr_task=$(( curr_task + 1 ))
+  echo "${curr_task}" > "${INFRA_DIR}/.curr_task"
+
+  # If we're done with all the tasks
+  if (( curr_task == TASKS_SIZE )); then
+    end_experiment
+
+    return 1
+  fi
+
+  # otherwise we check if we need to switch the task set and the treatment
+  # And go into the second half of the experiment
+  if (( curr_task == TASKS_SIZE / 2 + 1 )); then
+    echo "You have finished the first half of the experiment!"
+
+    # Updates the treatment and the task set for the user
+    if [[ "$(cat "${INFRA_DIR}/.treatment")" == "T" ]]; then
+      echo "N" > "${INFRA_DIR}/.treatment"
+    else
+      echo "T" > "${INFRA_DIR}/.treatment"
+    fi
+    if ((task_set == 1)); then
+      task_set=2
+    else
+      task_set=1
+    fi
+
+    print_treatment
+  fi
+
+  echo $(get_task_code) > "${INFRA_DIR}/.task_code"
+  SECONDS=0
+  time_elapsed=0
+  status="incomplete"
+
+  echo "start_task" > "${INFRA_DIR}/.command"
+
+  write_log
+  print_task
+}
+
+# Writes a command to the log file on the server with a POST request.
+write_log() {
+  curl -s -X POST ${SERVER_HOST}/${SERVER_ROUTE} \
+    -d user_id="$USER_ID" \
+    -d task_order="$(cat "${INFRA_DIR}/.task_order")" \
+    -d client_time_stamp="$(date --utc +%FT%TZ)" \
+    -d task_code="$(cat "${INFRA_DIR}/.task_code")" \
+    -d treatment="$(cat "${INFRA_DIR}/.treatment")" \
+    -d time_elapsed="${time_elapsed}" \
+    -d status="${status}" \
+    -d command="$(cat "${INFRA_DIR}/.command")" &> /dev/null
```

Comment:
Is there any chance this will fail?
If so, I suggest putting the output in a file, and if the status was failure, printing it -- at least during development and pilot studies.
If not, this is fine as is.

Reply:
This will fail under the following conditions:
1. If the user is not connected to the internet.
2. If the server is somehow down. This would require my CSE Homes to be down.
3. Internal server error. That is, the POST handler is buggy.
4. The `$SERVER_HOST` and/or `$SERVER_ROUTE` variables are incorrect.